cmd/mt-parrot/monitor.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"fmt"
+	"github.com/grafana/metrictank/stacktest/graphite"
+	"github.com/grafana/metrictank/stats"
+	log "github.com/sirupsen/logrus"
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+var (
+	httpError           = stats.NewCounter32("parrot.monitoring.error;error=http")
+	decodeError         = stats.NewCounter32("parrot.monitoring.error;error=decode")
+	parsePartitionError = stats.NewCounter32("parrot.monitoring.error;error=parsePartition")
+)
+
+type seriesStats struct {
+	lastTs uint32
+	//the partition currently being checked
+	nans int32
+	//the sum of abs(value - ts) across the time series
+	deltaSum float64
+	//the number of timestamps where value != ts
+	numNonMatching int32
+
+	//tracks the last seen non-NaN time stamp (useful for lag
+	lastSeen uint32
+}
+
+func monitor() {
+	for tick := range time.NewTicker(queryInterval).C {
+
+		query := graphite.ExecuteRenderQuery(buildRequest(tick))
+		if query.HTTPErr != nil {
+			httpError.Inc()
+		}
+		if query.DecodeErr != nil {
+			decodeError.Inc()
+		}
+
+		for _, s := range query.Decoded {
+			log.Infof("%d - %d", s.Datapoints[0].Ts, s.Datapoints[len(s.Datapoints)-1].Ts)
+			partition, err := strconv.Atoi(s.Target)
+			if err != nil {
+				log.Debug("unable to parse partition", err)
+				parsePartitionError.Inc()
+				continue
+			}
+			serStats := seriesStats{}
+			serStats.lastTs = s.Datapoints[len(s.Datapoints)-1].Ts
+
+			for _, dp := range s.Datapoints {
+
+				if math.IsNaN(dp.Val) {
+					serStats.nans += 1
+					continue
+				}
+				if diff := dp.Val - float64(dp.Ts); diff != 0 {
+					log.Debugf("partition=%d dp.Val=%f dp.Ts=%d diff=%f", partition, dp.Val, dp.Ts, diff)
+					serStats.lastSeen = dp.Ts
```

Comment:
i don't understand this. why do we want to track the "last seen" value so specifically for incorrect points, but don't update for correct points?

Reply:
Oops, that should be updated unless the value is `NaN` 401b2ca9990c754c7bd838dbcf618b06c9dcedcb